---
title: "Results from Metal Meta Meta; fixed effect"
author: "Alex Flynn-Carroll"
date: "2025-12-23"
output: html_document
---
###############################################################################################
# library
###############################################################################################

```{r packages}

library(tidyverse)
library(dplyr)
library(data.table)
library(ggplot2)
library(qqman)
library(PheWAS)
library(jpeg)
library(wesanderson)
library(ggrepel)
library(circlize)

```

###############################################################################################
# read in rolled up df
###############################################################################################

```{r}

# 1. List all files ending with "allchr.txt" in a folder
file_list <- list.files(path = "/mnt/arcus/lab/users/flynncarra/projects/20251222_AFR_EUR_meta/metal_results/meta_results/", pattern = "\\.txt$", full.names = TRUE)

```

```{r}

# 2. Read each file into a list of data frames
df_list <- lapply(file_list, function(f) {
  df <- read.table(f, header = TRUE, sep = "\t", stringsAsFactors = FALSE)
  df <- subset(df, P.value < 0.01)
  df$source_file <- basename(f)
  df
})

# 3. Combine all data frames into one
res <- do.call(rbind, df_list)

rm(df_list)

```

[1] 2733998       8

###############################################################################################
# filter df
###############################################################################################

#2056060
```{r}

dim(res)

res <- res %>%
  filter(!grepl("\\?", Direction))


```
# to 1490457


```{r}

res <- res %>%
  filter(nchar(Direction) == 2)

dim(res)
```

```{r}

#res_604 <- res[res$source_file != "meta_results_phe_604_1.txt",]

res <- res[res$source_file != "meta_results_phe_751.12_1.txt",]

#res <- res[res$source_file != "meta_results_phe_604_1.txt",]

#meta_results_phe_604_1.txt

```

```{r}

#female menstration
#res_626 <- res[res$source_file != "meta_results_phe_626_1.txt",]

#res <- res[res$source_file != "meta_results_phe_626_1.txt",]

#meta_results_phe_604_1.txt

```


```{r}

adhd_metal <- read.table("/mnt/arcus/lab/users/flynncarra/projects/20251222_AFR_EUR_meta/metal_results/meta_results_phe_313.1_1.txt", header = TRUE, sep = "\t", stringsAsFactors = FALSE)


```

###############################################################################################
# read in full GWAS with significant hits
###############################################################################################


###### get the full gws gwas
# added adhd for comparison to AFR
```{r}

# 1. List all files ending with "allchr.txt" in a folder
file_list_sig <- c("/mnt/arcus/lab/users/flynncarra/projects/20251222_AFR_EUR_meta/metal_results/meta_results/meta_results_phe_381.11_1.txt", "/mnt/arcus/lab/users/flynncarra/projects/20251222_AFR_EUR_meta/metal_results/meta_results/meta_results_phe_367.2_1.txt", "/mnt/arcus/lab/users/flynncarra/projects/20251222_AFR_EUR_meta/metal_results/meta_results/meta_results_phe_345.1_1.txt")

```

```{r}

# 2. Read each file into a list of data frames
df_list <- lapply(file_list_sig, function(f) {
  df <- read.table(f, header = TRUE, sep = "\t", stringsAsFactors = FALSE)
  df$source_file <- basename(f)
  df
})

# 3. Combine all data frames into one
res_sig <- do.call(rbind, df_list)

rm(df_list)

```

```{r only keep those in 2 or more arrays}

res_sig <- res_sig %>%
  filter(!grepl("\\?", Direction))

res_sig <- res_sig %>%
  filter(nchar(Direction) == 2)


```

```{r}

parts <- strsplit(as.character(res_sig$MarkerName), ":")
res_sig$CHR <- sapply(parts, `[`, 1)
res_sig$POS <- as.numeric(sapply(parts, `[`, 2))

```

```{r}

res_sig$CHR <- as.numeric(res_sig$CHR)

res_sig$POS <- as.numeric(res_sig$POS)

```

###############################################################################################
# Phecodes
###############################################################################################

```{r icd to phecode map}

phecode_map <- read.csv("../20250801_arcus_phenotype_counts/Phecode_map_v1_2_icd9_icd10cm.csv")

```

```{r unique phecode descriptions}

phecodes <- unique(phecode_map[,c("Phecode", "PhecodeString")])

phecodes <- phecodes %>%
  group_by(Phecode) %>%
  summarise(
    PhecodeString = paste(unique(PhecodeString), collapse = "; "),
    .groups = "drop"
  )

colnames(phecodes) <- c("phecode", "PhecodeString")

phecodes$phecode <- as.character(phecodes$phecode)

```

```{r}

res$phecode <- str_extract(res$source_file, "(?<=phe_)[0-9.]+(?=_1\\.txt)")

```

```{r}

top_hits <- res[res$P.value<=5e-8,]

top_hits <- left_join(top_hits, phecodes, by="phecode")

```

length(unique(res$MarkerName))
[1] 565009
> 0.05/565009
[1] 1.056334e-07
# number of variants used in the analysis

```{r}

write_csv(top_hits, "top_hits_afr_eur_1e6_meta_phewas.csv")

```

```{r}

filtered_phe_man_phe <- phewasManhattan(data.frame(phenotype = res$phecode, p = res$P.value), 
                                    title = "Arcus AFR-EUR Meta-Analyzed; Phecodes", 
                                    annotate.size=2,
                                    size.x.labels=5, 
                                    size.y.labels=5,
                                    significant.line = 5e-8,
                                    point.size =1)+ coord_cartesian(ylim = c(2, NA))

filtered_phe_man_phe

```

```{r}

# Save as JPEG, specify size in inches
jpeg(
  filename = "afr_eur_phewas_meta_gws_hwe_1e6 .jpg",
  width = 8,        # width in inches
  height = 4,       # height in inches
  units = "in",     # tells R that width/height are in inches
  res = 300         # resolution in DPI (dots per inch)
)

# Your plot
filtered_phe_man_phe

# Important: close the device to finalize the file
dev.off()

```

###############################################################################################
# Individual GWASs
###############################################################################################

```{r}

make_phecode_qq <- function(df, phecode_num, phecode_string, 
                            pval_col = "P.value",
                            outdir = ".") {
  # Extract p-values
  pvals <- df[[pval_col]]
  
  # Remove NA and invalid values
  pvals <- pvals[!is.na(pvals) & pvals > 0 & pvals <= 1]
  
  # Compute lambda
  chisq_vals <- qchisq(1 - pvals, df = 1)
  lambda_gc <- median(chisq_vals) / qchisq(0.5, df = 1)
  
  # Title for printing and plot
  title_text <- sprintf("%s (%s): A+E (Î» = %.3f)", 
                        phecode_string, phecode_num, lambda_gc)
  
  # Output filename
  file_name <- file.path(outdir, sprintf("QQ_%s_afr_eur.jpg", phecode_num))
  
  # Save QQ plot
  jpeg(file_name, width = 1200, height = 1200, res = 150)
  qq(pvals, main = title_text)
  dev.off()
  
  message("Saved: ", file_name)
  return(lambda_gc)
}

```

```{r}

phe_381.11_man <- manhattan(res_sig[res_sig$source_file == "meta_results_phe_381.11_1.txt",],
          chr = "CHR",
          bp = "POS",
          snp = "MarkerName",
          p = "P.value",
          main = "Suppurative and unspecified otitis media (381.11)",
          ylim = c(0, max(-log10(res_sig$P.value)) + 1))

phe_381.11_man

```

```{r}

jpeg("afr_eur_phe_381.11_man.jpg", width = 8*300, height = 6*300, res = 300)

manhattan(res_sig[res_sig$source_file == "meta_results_phe_381.11_1.txt",],
          chr = "CHR",
          bp = "POS",
          snp = "MarkerName",
          p = "P.value",
          main = "Suppurative and unspecified otitis media (381.11)",
          ylim = c(0, max(-log10(res_sig$P.value)) + 1))

dev.off()

```

```{r}

make_phecode_qq(res_sig[res_sig$source_file == "meta_results_phe_381.11_1.txt",], 381.11, "Suppurative and unspecified otitis media")

```

```{r}

jpeg("afr_eur_phe_367.2_man.jpg", width = 8*300, height = 6*300, res = 300)

manhattan(res_sig[res_sig$source_file == "meta_results_phe_367.2_1.txt",],
          chr = "CHR",
          bp = "POS",
          snp = "MarkerName",
          p = "P.value",
          main = "Astigmatism (367.2)",
          ylim = c(0, max(-log10(res_sig$P.value)) + 1))

dev.off()

```

```{r}

make_phecode_qq(res_sig[res_sig$source_file == "meta_results_phe_367.2_1.txt",], 367.2, "Astigmatism")

```

```{r}

jpeg("afr_eur_phe_345.1_man.jpg", width = 8*300, height = 6*300, res = 300)

manhattan(res_sig[res_sig$source_file == "meta_results_phe_345.1_1.txt",],
          chr = "CHR",
          bp = "POS",
          snp = "MarkerName",
          p = "P.value",
          main = "Epilepsy (345.1)",
          ylim = c(0, max(-log10(res_sig$P.value)) + 1))

dev.off()

```

```{r}

make_phecode_qq(res_sig[res_sig$source_file == "meta_results_phe_345.1_1.txt",], 345.1, "Epilepsy")

```