---
title: "AFR meta-phewas results"
author: "Alex Flynn-Carroll"
date: "2025-12-15"
output: html_document
---

###############################################################################################
# library
###############################################################################################

```{r packages}

library(tidyverse)
library(dplyr)
library(data.table)
library(ggplot2)
library(qqman)
library(PheWAS)
library(jpeg)
library(wesanderson)
library(ggrepel)
library(circlize)

```

###############################################################################################
# read in data
###############################################################################################

```{r}

# 1. List all files ending with "allchr.txt" in a folder
file_list <- list.files(path = "/mnt/arcus/lab/users/flynncarra/projects/20251215_afr_meta/metal_results/", pattern = "\\.txt$", full.names = TRUE)

```

```{r}

# 2. Read each file into a list of data frames
df_list <- lapply(file_list, function(f) {
  df <- read.table(f, header = TRUE, sep = "\t", stringsAsFactors = FALSE)
  df <- subset(df, P.value < 0.01)
  df$source_file <- basename(f)
  df
})

# 3. Combine all data frames into one
res <- do.call(rbind, df_list)

rm(df_list)

```

```{r icd to phecode map}

phecode_map <- read.csv("../20250801_arcus_phenotype_counts/Phecode_map_v1_2_icd9_icd10cm.csv")

```

###############################################################################################
# filter df
###############################################################################################

```{r}

dim(res)

res <- res %>%
  filter(str_count(Direction, fixed("?")) <= 3)

```

```{r}

res <- res[res$source_file != "meta_results_phe_751.12_1.txt",]

#res <- res[res$source_file != "meta_results_phe_604_1.txt",]

#meta_results_phe_604_1.txt

```

```{r}

res$phecode <- str_extract(res$source_file, "(?<=phe_)[0-9.]+(?=_1\\.txt)")

```

###############################################################################################
# Phecodes
###############################################################################################

```{r unique phecode descriptions}

phecodes <- unique(phecode_map[,c("Phecode", "PhecodeString")])

phecodes <- phecodes %>%
  group_by(Phecode) %>%
  summarise(
    PhecodeString = paste(unique(PhecodeString), collapse = "; "),
    .groups = "drop"
  )

colnames(phecodes) <- c("phecode", "PhecodeString")

phecodes$phecode <- as.character(phecodes$phecode)

```

```{r}

top_hits <- res[res$P.value<=1.05654e-07,]

top_hits <- left_join(top_hits, phecodes, by="phecode")

```

> length(unique(res$MarkerName))
[1] 473243
> 0.05/473243
[1] 1.05654e-07
# number of variants used in the anla

```{r}

write_csv(top_hits, "top_hits_afr_meta_phewas.csv")

```

```{r}

filtered_phe_man_phe <- phewasManhattan(data.frame(phenotype = res$phecode, p = res$P.value), 
                                    title = "Arcus AFR Meta-Analyzed; Phecodes", 
                                    annotate.size=2,
                                    size.x.labels=5, 
                                    size.y.labels=5,
                                    significant.line = 5e-8,
                                    point.size =1)+ coord_cartesian(ylim = c(2, NA))

filtered_phe_man_phe

```

```{r}

# Save as JPEG, specify size in inches
jpeg(
  filename = "saige_phewas_meta_gws_afr_5e8.jpg",
  width = 8,        # width in inches
  height = 4,       # height in inches
  units = "in",     # tells R that width/height are in inches
  res = 300         # resolution in DPI (dots per inch)
)

# Your plot
filtered_phe_man_phe

# Important: close the device to finalize the file
dev.off()

```

###############################################################################################
# individuals GWAS prep
###############################################################################################

###### get the full gws gwas
```{r}

gws_phe <- unique(top_hits$source_file)

gws_paths <- file.path(
  "/mnt/arcus/lab/users/flynncarra/projects/20251215_afr_meta/metal_results",
  gws_phe
)

```

```{r}

# 2. Read each file into a list of data frames
df_list <- lapply(gws_paths, function(f) {
  df <- read.table(f, header = TRUE, sep = "\t", stringsAsFactors = FALSE)
  df$source_file <- basename(f)
  df
})

# 3. Combine all data frames into one
res_sig <- do.call(rbind, df_list)

rm(df_list)

```

```{r only keep those in 2 or more arrays}

res_sig <- res_sig %>%
  filter(str_count(Direction, fixed("?")) <= 3)

```

```{r}

parts <- strsplit(as.character(res_sig$MarkerName), ":")
res_sig$CHR <- sapply(parts, `[`, 1)
res_sig$POS <- as.numeric(sapply(parts, `[`, 2))

```

```{r}

res_sig$CHR <- as.numeric(res_sig$CHR)

res_sig$POS <- as.numeric(res_sig$POS)

```

###############################################################################################
# plot individual GWAS
###############################################################################################

```{r}

make_phecode_qq <- function(df, phecode_num, phecode_string, 
                            pval_col = "P.value",
                            outdir = ".") {
  # Extract p-values
  pvals <- df[[pval_col]]
  
  # Remove NA and invalid values
  pvals <- pvals[!is.na(pvals) & pvals > 0 & pvals <= 1]
  
  # Compute lambda
  chisq_vals <- qchisq(1 - pvals, df = 1)
  lambda_gc <- median(chisq_vals) / qchisq(0.5, df = 1)
  
  # Title for printing and plot
  title_text <- sprintf("%s (%s): AFR (Î» = %.3f)", 
                        phecode_string, phecode_num, lambda_gc)
  
  # Output filename
  file_name <- file.path(outdir, sprintf("QQ_%s_afr.jpg", phecode_num))
  
  # Save QQ plot
  jpeg(file_name, width = 1200, height = 1200, res = 150)
  qq(pvals, main = title_text)
  dev.off()
  
  message("Saved: ", file_name)
  return(lambda_gc)
}

```

```{r}

phe_282_man <- manhattan(res_sig[res_sig$source_file == "meta_results_phe_282_1.txt",],
          chr = "CHR",
          bp = "POS",
          snp = "MarkerName",
          p = "P.value",
          main = "AFR: Hereditary hemolytic anemias (282)",
          ylim = c(0, max(-log10(res_sig$P.value)) + 1))

phe_282_man

```
```{r}

jpeg("meta_man_phe_282_afr.jpg", width = 8*300, height = 6*300, res = 300)

manhattan(res_sig[res_sig$source_file == "meta_results_phe_282_1.txt",],
          chr = "CHR",
          bp = "POS",
          snp = "MarkerName",
          p = "P.value",
          main = "AFR: Hereditary hemolytic anemias (282)",
          ylim = c(0, max(-log10(res_sig$P.value)) + 1))

dev.off()

```

```{r}

make_phecode_qq(res_sig[res_sig$source_file == "meta_results_phe_282_1.txt",], 282, "Hereditary hemolytic anemias")

```


```{r}

phe_261.4_man <- manhattan(res_sig[res_sig$source_file == "meta_results_phe_261.4_1.txt",],
          chr = "CHR",
          bp = "POS",
          snp = "MarkerName",
          p = "P.value",
          main = "AFR: Vitamin D deficiency (261.4)",
          ylim = c(0, max(-log10(res_sig$P.value)) + 1))

phe_261.4_man

```
```{r}

jpeg("meta_man_phe_261.4_afr.jpg", width = 8*300, height = 6*300, res = 300)

manhattan(res_sig[res_sig$source_file == "meta_results_phe_261.4_1.txt",],
          chr = "CHR",
          bp = "POS",
          snp = "MarkerName",
          p = "P.value",
          main = "AFR: Vitamin D deficiency (261.4)",
          ylim = c(0, max(-log10(res_sig$P.value)) + 1))

dev.off()

```

```{r}

make_phecode_qq(res_sig[res_sig$source_file == "meta_results_phe_261.4_1.txt",], 261.4, "Vitamin D deficiency")

```

```{r}

phe_79_man <- manhattan(res_sig[res_sig$source_file == "meta_results_phe_79_1.txt",],
          chr = "CHR",
          bp = "POS",
          snp = "MarkerName",
          p = "P.value",
          main = "AFR: Viral infection (79)",
          ylim = c(0, max(-log10(res_sig$P.value)) + 1))

phe_79_man

```

```{r}

phe_274.2_man <- manhattan(res_sig[res_sig$source_file == "meta_results_phe_474.2_1.txt",],
          chr = "CHR",
          bp = "POS",
          snp = "MarkerName",
          p = "P.value",
          main = "AFR: Chronic tonsillitis and adenoiditis (274.2)",
          ylim = c(0, max(-log10(res_sig$P.value)) + 1))

phe_274.2_man

```
```{r}

jpeg("meta_man_phe_474.2_afr.jpg", width = 8*300, height = 6*300, res = 300)

manhattan(res_sig[res_sig$source_file == "meta_results_phe_474.2_1.txt",],
          chr = "CHR",
          bp = "POS",
          snp = "MarkerName",
          p = "P.value",
          main = "AFR: Chronic tonsillitis and adenoiditis (274.2)",
          ylim = c(0, max(-log10(res_sig$P.value)) + 1))

dev.off()

```

```{r}

make_phecode_qq(res_sig[res_sig$source_file == "meta_results_phe_474.2_1.txt",], 474.2, "Chronic tonsillitis and adenoiditis")

```

```{r}

phe_350.2_man <- manhattan(res_sig[res_sig$source_file == "meta_results_phe_350.2_1.txt",],
          chr = "CHR",
          bp = "POS",
          snp = "MarkerName",
          p = "P.value",
          main = "AFR: Abnormality of gait (350.2)",
          ylim = c(0, max(-log10(res_sig$P.value)) + 1))

phe_350.2_man

```

```{r}

phe_313.1_man <- manhattan(res_sig[res_sig$source_file == "meta_results_phe_313.1_1.txt",],
          chr = "CHR",
          bp = "POS",
          snp = "MarkerName",
          p = "P.value",
          main = "AFR: Attention deficit hyperactivity disorder (313.1)",
          ylim = c(0, max(-log10(res_sig$P.value)) + 1))

phe_313.1_man

```

```{r}

jpeg("meta_man_phe_313.1_afr.jpg", width = 8*300, height = 6*300, res = 300)

manhattan(res_sig[res_sig$source_file == "meta_results_phe_313.1_1.txt",],
          chr = "CHR",
          bp = "POS",
          snp = "MarkerName",
          p = "P.value",
          main = "AFR: Attention deficit hyperactivity disorder (313.1)",
          ylim = c(0, max(-log10(res_sig$P.value)) + 1))


dev.off()

```

```{r}

make_phecode_qq(res_sig[res_sig$source_file == "meta_results_phe_313.1_1.txt",], 313.1, "Attention deficit hyperactivity disorder")

```

```{r}

phe_345_man <- manhattan(res_sig[res_sig$source_file == "meta_results_phe_345_1.txt",],
          chr = "CHR",
          bp = "POS",
          snp = "MarkerName",
          p = "P.value",
          main = "AFR: Epilepsy, recurrent seizures, convulsions (345)",
          ylim = c(0, max(-log10(res_sig$P.value)) + 1))

phe_345_man


```
```{r}

jpeg("meta_man_phe_345_afr.jpg", width = 8*300, height = 6*300, res = 300)

manhattan(res_sig[res_sig$source_file == "meta_results_phe_345_1.txt",],
          chr = "CHR",
          bp = "POS",
          snp = "MarkerName",
          p = "P.value",
          main = "AFR: Epilepsy, recurrent seizures, convulsions (345)",
          ylim = c(0, max(-log10(res_sig$P.value)) + 1))


dev.off()

```

```{r}

make_phecode_qq(res_sig[res_sig$source_file == "meta_results_phe_345_1.txt",], 345, "Epilepsy, recurrent seizures, convulsions")

```

276.5

```{r}

phe_276.5_man <- manhattan(res_sig[res_sig$source_file == "meta_results_phe_276.5_1.txt",],
          chr = "CHR",
          bp = "POS",
          snp = "MarkerName",
          p = "P.value",
          main = "AFR: Hypovolemia (276.5)",
          ylim = c(0, max(-log10(res_sig$P.value)) + 1))

phe_276.5_man


```
```{r}

jpeg("meta_man_phe_276.5_afr.jpg", width = 8*300, height = 6*300, res = 300)

manhattan(res_sig[res_sig$source_file == "meta_results_phe_276.5_1.txt",],
          chr = "CHR",
          bp = "POS",
          snp = "MarkerName",
          p = "P.value",
          main = "AFR: Hypovolemia (276.5)",
          ylim = c(0, max(-log10(res_sig$P.value)) + 1))

dev.off()

```

```{r}

make_phecode_qq(res_sig[res_sig$source_file == "meta_results_phe_276.5_1.txt",], 276.5, "Hypovolemia")

```

###############################################################################################
# Phecode Map
###############################################################################################

```{r}
top_hits <- res[res$P.value<=1.05654e-07,]
top_hits <- addPhecodeInfo(top_hits)
```

```{r}

res <- addPhecodeInfo(res)

```


```{r}

phecode_plot <- unique(phecode_map[,c("Phecode","PhecodeString","PhecodeCategory")])

```

```{r}


category_order <- c(
  "circulatory system",
  "congenital anomalies",
  "dermatologic",
  "digestive",
  "endocrine/metabolic",
  "genitourinary",
  "hematopoietic",
  "infectious diseases",
  "injuries & poisonings",
  "mental disorders",
  "musculoskeletal",
#  "neoplasms",
  "neurological",
#  "other",
  "pregnancy complications",
  "respiratory",
  "sense organs",
  "symptoms"
)

res <- res %>%
  mutate(
    category = factor(group, levels = category_order),
    neglog10p = -log10(P.value),
    cat_index = as.numeric(category)
  )


```

```{r}

set.seed(1)
res <- res %>%
  mutate(
    x_jitter = cat_index + runif(n(), -0.3, 0.3)
  )

alpha <- 0.05
bonf_thresh <- -log10(5e-8)

```

```{r}

# Create mapping
cat_labels <- res %>%
  dplyr::select(cat_index, category) %>%
  dplyr::distinct() %>%
  dplyr::arrange(cat_index)

```

```{r}
gw_thresh <- 5e-8

# Filter only significant hits
sig_hits <- res %>%
  filter(P.value < gw_thresh)

# Select top hit per phenotype among significant hits
top_hits <- sig_hits %>%
  group_by(phecode) %>%
  slice_max(order_by = neglog10p, n = 1, with_ties = FALSE) %>%
  ungroup()

```

```{r}

phewas_plot1 <- ggplot(res, aes(x = x_jitter, y = neglog10p, color = group)) +
  geom_point(alpha = 0.7, size = 2) +
  geom_hline(yintercept = bonf_thresh, linetype = "dashed", color = "black") +
  geom_text_repel(
    data = top_hits,                # only top hit per phenotype
    aes(label = description),
    size = 2,
    max.overlaps = 30,
    box.padding = 0.4,
    color = "black"
  ) +
  scale_x_continuous(
    breaks = cat_labels$cat_index,
    labels = cat_labels$category,
    expand = expansion(mult = c(0.02, 0.02))
  ) +
  labs(x = NULL, y = expression(-log[10](italic(P))), title = "AFR Phecode PheWAS") +
  theme_bw() +
  theme(
    axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, color = "black"),
    axis.title = element_text(color = "black"),
    plot.title = element_text(color = "black"),
    legend.position = "none",
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank()
  )


```

```{r}

# Save as JPEG, specify size in inches
jpeg(
  filename = "saige_phewas_meta_gws_afr_5e8_v2.jpg",
  width = 8,        # width in inches
  height = 4,       # height in inches
  units = "in",     # tells R that width/height are in inches
  res = 300         # resolution in DPI (dots per inch)
)

# Your plot
phewas_plot1

# Important: close the device to finalize the file
dev.off()

```

```{r}


# Then in ggplot
 ggplot(res, aes(x = x_jitter, y = neglog10p, color = group)) +
  geom_point(alpha = 0.7, size = 2) +
  geom_hline(yintercept = bonf_thresh, linetype = "dashed", color = "black") +
  geom_text_repel(
    data = subset(res, neglog10p > bonf_thresh),
    aes(label = description),
    size = 3.5,
    max.overlaps = 30,
    box.padding = 0.4,
    color = "black"  
  ) +
  scale_x_continuous(
    breaks = cat_labels$cat_index,
    labels = cat_labels$category,
    expand = expansion(mult = c(0.02, 0.02))
  ) +
  labs(x = NULL, y = expression(-log[10](italic(P))), title = "PheWAS plot") +
  theme_bw() +
  theme(
    axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1),
    legend.position = "none",
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank()
  )

```